### ISSUE TITLE
changed-tiles detection when OSM objects are moved to other tiles

### ISSUE THREAD

**Issue by ImreSamu** [osm2vectortiles/osm2vectortiles#210]
I have found a strange behavior in the change tiles detection part. 

### Test:
WHEN 1 POI moved to other tiles
* FROM [ lat="33.3" lon="33.3"]   tile:  [`14/9707/6583 `] 
* TO       [ lat="3.14" lon="3.14"]   tile: [`14/8334/8049`] 

THEN my expectation is that  changed tiles  file  (  ` ./export/tiles.txt` ) should contain
* the old tile [`14/9707/6583 `]      -  for invalidate the POI
* and the new  [`14/8334/8049`] tile. 

BUT  now the old tile [`14/9707/6583 `] is missing  (  ` ./export/tiles.txt` )

So in extreme case  ( "neighbour tiles" ) this POI will be show twice on the map.

### My test method:

* tested  version  ( current master ): 
 * `commit bb256f5b6a7b1f3d1e99b0ca72b4aa5ad5ec5e08` 
 * `Date:   Fri Apr 1 23:05:19 2016 +0200`
* Removed all osm2vectortiles/*  docker images  & rebuilded with  `make fast`

Simple test code: 

```bash
#!/bin/bash
set -o errexit
set -o pipefail
set -o nounset

# clean start
docker-compose stop
docker-compose rm -f -v
rm -f ./import/*

docker-compose up -d postgis
sleep 10
 
# create m0.osm     
cat <<EOT > m0.osm
<?xml version="1.0" encoding="UTF-8"?>
 <osm version="0.6" generator="osm2vectortiles testdata" upload="false">
 <node id="3" changeset="1" timestamp="2016-01-01T00:00:00Z" 
              version="1" visible="true" user="o2v" uid="1" lat="33.3" lon="33.3">
 <tag k="name" v="OSM2Vectortiles"/>
 <tag k="shop" v="supermarket"/>
 </node>
 </osm>
EOT

# create m1.osm :    only : "lat","lon"  modified
cat <<EOT > m1.osm 
<?xml version="1.0" encoding="UTF-8"?>
 <osm version="0.6" generator="osm2vectortiles testdata" upload="false">
 <node id="3" changeset="2" timestamp="2016-03-01T00:00:00Z" 
              version="2" visible="true" user="o2v" uid="1" lat="3.14" lon="3.14">
 <tag k="name" v="OSM2Vectortiles"/>
 <tag k="shop" v="supermarket"/>
 </node>
 </osm>
EOT
 

# create diff files  
osmconvert m0.osm                 --timestamp=2016-01-01T00:00:00Z -o=./import/m0.pbf
osmconvert m0.osm  m1.osm  --diff --timestamp=2016-03-03T00:00:00Z -o=mdiff.osc
cat mdiff.osc | gzip  > ./import/latest.osc.gz

#Step import
docker-compose run import-osm
docker-compose run import-sql
docker-compose run changed-tiles
cp ./export/tiles.txt  ./export/tiles_01_import.txt


#Step diff
docker-compose run import-osm-diff
docker-compose run changed-tiles
cp ./export/tiles.txt  ./export/tiles_02_diff.txt


echo "Step import - changed tiles"
cat ./export/tiles_01_import.txt

echo "Step diff - changed tiles"
cat ./export/tiles_02_diff.txt


# ----- Now ------
# Step import - changed tiles
# 14/9707/6583
# Step diff - changed tiles
# 14/8334/8049

# ---- Expecting -----
# Step import - changed tiles
# 14/9707/6583
# Step diff - changed tiles
# 14/8334/8049
# 14/9707/6583         < --- Missing! 

```

related to  ( https://github.com/osm2vectortiles/osm2vectortiles/issues/183 )

**Comment by lukasmartinelli** [comment]
First of all, thanks for the great issue (including test!!) and the [OSM talk on osm2vectortiles today](http://www.meetup.com/OpenStreetMap-Hungary/events/229839021/) :grin: 

We need to enable delete tracking for updates as well.
Your test works now with the 21a9173 commit.
I hope this does not slow down the import as crazy as last time - because this time we only use the delete trigger. Will need to test it on a reasonably large extract or import the world again.

**Comment by lukasmartinelli** [comment]
Yess!! Doesn't slow down the diff import as much as I feared. Diff import of ~15 days only takes **6h57m** even though there is a trigger on every `DELETE` call.

For ~15 days back it recorded **2'839'020** updated or deleted geometries that were tracked in the `osm_delete` table.

That is really good news. So as lesson learnt: Complexity in triggers matters **a lot!** If the trigger is simple enough performance can still be ok.

**Comment by lukasmartinelli** [comment]
Okay but as downside the changed tiles detection is now running since 14 hrs instead of one.

But I hope we can fix this by introducing a GIST index to the deleted geometry table. The recursive algorithm for detecting the dirty tiles makes heavy use `geom1 && geom2`.

**Comment by ImreSamu** [comment]
optimalisation is not easy  :) 

>changed tiles detection is now running since 14 hrs instead of one.

imho:  it is not parallel, or can be more parallel  

my tips:
*  forcing an extreme parallelism in the postprocessing part  (  by `imposm3` tables ) 
 *   now only 1 `osm_delete`  table  ->  why not for every `imposm3` tables? my quess: it can be a bottleneck,  because of the parallel insert.
    * osm_admin_linestring_delete
    * osm_aero_linestring_delete
    * osm_aero_polygon_delete
    * ...
   * using  'unlogged' or 'temp'    *_delete tables  
     * CREATE UNLOGGED TABLE  osm_admin_linestring_delete AS ...
   * parallel generation of  changed tiles
     *  ->  ./export/admin_z2toz6_tiles.txt 
     *  ->  ./export/admin_z7toz14_tiles.txt    
     *  ->  ./export/aeroway_z12toz14_tiles.txt
     *  ....

     and at the end   -  fast linux postprocessing  ( or using a `process  queue` .. )
     *  cat ./export/admin_z2toz6_tiles.txt  ./export/admin_z7toz14_tiles.txt  ... | sort -u > ./export/tiles.txt     


> changed tiles detection is now running since 14 hrs instead of one.

with more optimalisation we can create an intelligent  `changed tile processing queue` 
like ..
* if the ./export/admin_z2toz6_tiles.txt   generation is the fastest ( 30 min)  then we can add to the `processing queue`  for generating tiles 
*  later the ./export/admin_z7toz14_tiles.txt  is finished , so we can add this tiles for processing ( but adding only the new tiles!    
similar than :  diff  `admin_z2toz6_tiles.txt`  `admin_z7toz14_tiles.txt`   )

**Comment by lukasmartinelli** [comment]
First of all, thanks again for the very insightful points raised and like always great suggestions @ImreSamu 

> But I hope we can fix this by introducing a GIST index to the deleted geometry table. The recursive algorithm for detecting the dirty tiles makes heavy use geom1 && geom2.

Yes the missing GIST index was the problem - now calculating the dirty tiles for just the deletes is done in 30minutes. So good news there.

> forcing an extreme parallelism in the postprocessing part ( by imposm3 tables )

Writing to the `delete` tables is not the problem in my experience but the cost of just triggering the trigger for each delete. In the main *imposm3* import we write sometimes half a billion records into a table and it scales.

> parallel generation of changed tiles

In the previous solution on NodeJS we did 15 parallel SQL queries and then merged the results back together. Then we switched to doing just one SQL query (which in turn does a `UNION` of 15 SQL queries) and found out it was just as fast - main cost is simply waiting for all that massive IO that is happening.

> with more optimalisation we can create an intelligent

Yes not having to wait for all tiles to be calculated means we can get started with rendering quicker.
But by waiting for all tiles to be calculated removing the duplicate tiles is easier.
If we can get the changed tiles calculation back to ~1 hour we can probably leave it like this with one big calculation of the tiles and then starting to render.

If it takes longer I want to try the approach getting faster tile feedback and already start rendering the dirty tiles.
